platform/pulp/platform/tasks/task_system.py

```diff
@@ -1,4 +1,548 @@
+    tasks.
+    """
+    pass
+
+
+@task(base=PulpTask, acks_late=True)
+def _queue_reserved_task(name, task_id, resource_id, inner_args, inner_kwargs):
+    """
+    A task that encapsulates another task to be dispatched later. This task being encapsulated is
+    called the "inner" task, and a task name, UUID, and accepts a list of positional args
+    and keyword args for the inner task. These arguments are named inner_args and inner_kwargs.
+    inner_args is a list, and inner_kwargs is a dictionary passed to the inner task as positional
+    and keyword arguments using the * and ** operators.
+
+    The inner task is dispatched into a dedicated queue for a worker that is decided at dispatch
+    time. The logic deciding which queue receives a task is controlled through the
+    find_worker function.
+
+    :param name:          The name of the task to be called
+    :type name:           basestring
+    :param inner_task_id: The UUID to be set on the task being called. By providing
+                          the UUID, the caller can have an asynchronous reference to the inner task
+                          that will be dispatched.
+    :type inner_task_id:  basestring
+    :param resource_id:   The name of the resource you wish to reserve for your task. The system
+                          will ensure that no other tasks that want that same reservation will run
+                          concurrently with yours.
+    :type  resource_id:   basestring
+
+    :return: None
+    """
+    while True:
+        # Find a worker who already has this reservation, it is safe to send this work to them
+        try:
+            worker = ReservedResource.objects.get(resource=resource_id).worker
+        except ReservedResource.DoesNotExist:
+            pass
+        else:
+            break
+
+        try:
+            worker = _get_unreserved_worker()
+        except Worker.DoesNotExist:
+            pass
+        else:
+            break
+
+        # No worker is ready for this work, so we need to wait
+        time.sleep(0.25)
+
+    task_status = TaskStatus.objects.get(id=task_id)
+    ReservedResource(task=task_status, worker=worker, resource=resource_id).save()
+
+    inner_kwargs['routing_key'] = worker.name
+    inner_kwargs['exchange'] = DEDICATED_QUEUE_EXCHANGE
+    inner_kwargs['task_id'] = task_id
+
+    try:
+        celery.tasks[name].apply_async(*inner_args, **inner_kwargs)
+    finally:
+        _release_resource.apply_async((task_id, ), routing_key=worker.name,
+                                      exchange=DEDICATED_QUEUE_EXCHANGE)
+
+
+def _is_worker(worker_name):
+    """
+    Strip out workers that should never be assigned work. We need to check
+    via "startswith()" since we do not know which host the worker is running on.
+    """
+
+    if worker_name.startswith(SCHEDULER_WORKER_NAME) or \
+       worker_name.startswith(RESOURCE_MANAGER_QUEUE):
+        return False
+    return True
+
+
+def _get_unreserved_worker():
```

Comment:
I am not sure if you forgot or did not mean to, on purpose, but i don't see `get_worker_for_reservation()`